x/mongo/driver/crypt.go

```diff
@@ -43,24 +43,46 @@ type CryptOptions struct {
 	BypassAutoEncryption bool
 }
 
-// Crypt consumes the libmongocrypt.MongoCrypt type to iterate the mongocrypt state machine and perform encryption
+// Crypt is an interface implemented by types that can encrypt and decrypt instances of
+// bsoncore.Document.
+//
+// Users should rely on the driver's crypt type (used by default) for encryption and decryption
+// unless they are perfectly confident in another implementation of Crypt.
+type Crypt interface {
+	// Encrypt encrypts the given command.
+	Encrypt(ctx context.Context, db string, cmd bsoncore.Document) (bsoncore.Document, error)
+	// Decrypt decrypts the given command response.
+	Decrypt(ctx context.Context, cmdResponse bsoncore.Document) (bsoncore.Document, error)
+	// CreateDataKey creates a data key using the given KMS provider and options.
+	CreateDataKey(ctx context.Context, kmsProvider string, opts *options.DataKeyOptions) (bsoncore.Document, error)
+	// EncryptExplicit encrypts the given value with the given options.
+	EncryptExplicit(ctx context.Context, val bsoncore.Value, opts *options.ExplicitEncryptionOptions) (byte, []byte, error)
+	// DecryptExplicit decrypts the given encrypted value.
+	DecryptExplicit(ctx context.Context, subtype byte, data []byte) (bsoncore.Value, error)
+	// Close cleans up any resources associated with the Crypt instance.
+	Close()
+	// BypassAutoEncryption returns true if auto-encryption should be bypassed.
+	BypassAutoEncryption() bool
+}
+
+// crypt consumes the libmongocrypt.MongoCrypt type to iterate the mongocrypt state machine and perform encryption
 // and decryption.
-type Crypt struct {
+type crypt struct {
 	mongoCrypt *mongocrypt.MongoCrypt
 	collInfoFn CollectionInfoFn
 	keyFn      KeyRetrieverFn
 	markFn     MarkCommandFn
 
-	BypassAutoEncryption bool
+	bypassAutoEncryption bool
 }
 
 // NewCrypt creates a new Crypt instance configured with the given AutoEncryptionOptions.
-func NewCrypt(opts *CryptOptions) (*Crypt, error) {
-	c := &Crypt{
+func NewCrypt(opts *CryptOptions) (Crypt, error) {
```

Comment:
I see that according to the `crypt` struct that one of the fields is a MongoCrypt but I'm a bit confused about the difference between the two. After looking at mongocrypt.go, it seems like MongoCrypt creates "contexts" for different encryption/decryption commands -- what exactly are "contexts" and how are they used?

Reply:
Great question(s). 

Our implementation of CSFLE in the Go driver relies on a separate CSFLE implementation in C called [libmongocrypt](https://github.com/mongodb/libmongocrypt). This is the case for quite a few drivers. A lot of our objects and methods associated with client-side-encryption wrap this C library using a package called [Cgo](https://pkg.go.dev/cmd/cgo), a really cool, standard interop library. 

The `mongoCrypt` field of the `crypt` struct is actually a Go wrapper around the C type `mongocrypt_t` from `libmongocrypt`. Those "contexts" you're seeing in `mongocrypt.go` are also Go wrappers around C types. In this case, we're [wrapping](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/mongocrypt/mongocrypt_context.go#L17-L20) the `mongocrypt_ctx_t` `libmongocrypt` type defined [here](https://github.com/mongodb/libmongocrypt/blob/master/src/mongocrypt-ctx-private.h#L64-L75). I _think_ that type just contains meta-information about the current encryption/decryption, but @kevinAlbs would definitely know better.